### ISSUE TITLE
Problem with ZIndex 

### ISSUE THREAD

**Issue by sylvainbeo** [boundlessgeo/ol-mapbox-style#119]
Hi,

I've noticed big changes in version 3.3.0.
I was using version 3.2, and after migrating to 3.3.0 i think there is an issue with the lazer's zindex. In previous version, ZIndex was set  direcly in the layer. Now it is only set in the processStyle() function. I've tried to debug, and my code never goes there (i use apply and not olms). Zindex of the layer seems to be corrupted then, and my background is overlaying my other layers.

Thank you.

**Comment by ahocevar** [comment]
Can you share your code and the style object you are using? `processStyle` should always be called, because that's the function that creates the layers.

**Comment by sylvainbeo** [comment]
@ahocevar Just like that :

```
...
 this.map = new Map({
            layers: [],
            target: 'map',
            overlays: [this.popup],
            view: new View({
                center: transform(center, configData.data_epsg, configData.map_epsg),
                zoom: zoom,
                extent: configData.map_extent,
                minZoom: configData.map_min_zoom
            })
        });

// Add OSM vector tile layer
apply(this.map, "http://osm.oslandia.io/styles/klokantech-basic/style.json");
...
```

**Comment by ahocevar** [comment]
@sylvainbeo I'm not sure I understand. The style you apply adds a single layer to the map, with zIndex 41. The map does not contain any other layers, and the background is applied to the map's viewport div. Can you provide more detail on how to see the issue?

**Comment by sylvainbeo** [comment]
>  The style you apply adds a single layer to the map

Yes, this is the background. Later in the code i add other layers dynamically (and force the zIndex to specific orders)

> with zIndex 41

Woa ? Where did you see that ?

**Comment by ahocevar** [comment]
> Woa ? Where did you see that ?

When you have a reference to the map, you can do something like
```js
map.getLayers().item(0).getZIndex();
```
The z-index in the current version is the last layer's array index from the `layers` array of the Mapbox Style object that the OpenLayers layer contains. Previous versions of the library had a lower z-index (i.e. the index of the first layer that the OpenLayers layer contains). Future versions (after e5cb7c8e579b132decd5d6d4c5fe3d51b9c2a168) will have the lower z-index again.

**Comment by sylvainbeo** [comment]
Mmm, i'm not sure to well understand. For the current version, each layer aaded with the apply() function will have a zIndex of 41 if it is the version layer added to the map ? 

If so, that's weird, but ok.

That means, i have to set the zIndex of my others layers above 41.

I know the getZIndex() function. But problem is that apply() is asynchronous, so i can not get the zIndex of the layer right after calling the function. I tried with a promise, but did not work either.

**Comment by ahocevar** [comment]
The z-index of a layer added with `apply` or the library's default export depends on the style object, and is definitely not always 41. To get a safe z-index for your additional layers, you can do something like this:
```js
apply(map, 'path/to/my/style.json');
const zIndex = map.get('mapbox-style').layers.length;
myLayer.setZIndex(zIndex++);
map.addLayer(myLayer); 
```

Or you can check the number of layers in the style object's layers array and give your other layers a higher zIndex:
```js
const zIndex = glStyle.layers.length;
myLayer.setZIndex(zIndex);
```

**Comment by ahocevar** [comment]
I think we'll change the behavior to not apply any z-index to the layers created by this library. Instead, we'll just make sure they're added to the map in the correct order. That should make it easier to use for users that add their own layers in addition to the ones created by the library.

**Comment by sylvainbeo** [comment]
@ahocevar Ok, that's sounds good to me. Thank you.
For the moment, i don't use the lastest version of the lib. 
I can not apply the code you suggested above, does not work in my case.

I thought apply() was asynchronous, if so, we can't get back the zIndex just after the call.

**Comment by ahocevar** [comment]
@sylvainbeo Note that the API of the default export is the same as the one for `apply`, with the only difference that it returns a promise instead of a map.

**Comment by ahocevar** [comment]
Published v3.4.0 with the changes from #121.
